src/main/java/frc/team2412/robot/Subsystems/IndexerSubsystem.java

```diff
@@ -0,0 +1,257 @@
+
+public class IndexerSubsystem extends SubsystemBase {
+	public CANSparkMax indexBackMotor, indexFrontMotor, indexMidMotor;
+	public SpeedControllerGroup indexMotors, indexSideMotors;
+
+	public DigitalInput back, backMid, mid, frontMid, front, intakeFront, intakeBack;
+
+	// NUMBER OF BALLS IN SYSTEM
+	public int numBalls = 0;
+
+	// SIDE OF INDEXER WITH LESS BALLS
+	public IntakeDirection ballUnbalancedSide;
+
+	// to run function once until reset
+	private int run = 0;
+
+	public IndexerSubsystem() {
+		// Motors
+		indexBackMotor = RobotMap.indexFrontMotor;
+		indexFrontMotor = RobotMap.indexBackMotor;
+		indexMidMotor = RobotMap.indexMidMotor;
+		indexMotors = new SpeedControllerGroup(indexFrontMotor, indexBackMotor, indexMidMotor);
+		indexSideMotors = new SpeedControllerGroup(indexFrontMotor, indexBackMotor);
+
+		// Sensors
+		back = RobotMap.back;
+		backMid = RobotMap.backMid;
+		mid = RobotMap.mid;
+		frontMid = RobotMap.frontMid;
+		front = RobotMap.front;
+
+		intakeFront = RobotMap.intakeFront;
+		intakeBack = RobotMap.intakeBack;
+
+		setDefaultCommand(new ProcessBalls(this, new ShootBalls(this), new IntakeBalls(this), new SwitchBalls(this)));
+	}
+
+	// shifters
+
+	// Motors
+	public void stopAll() {
+		indexMotors.set(0.0);
+	}
+
+	public void stopSides() {
+		indexSideMotors.set(0.0);
+	}
+
+	// Sensors
+	public boolean allSensorsOn() {
+		if (back.get() && backMid.get() && mid.get() && frontMid.get() && front.get()) {
+			return true;
+		}
+
+		return false;
+		// check if all sensors are on
+	}
+
+	public boolean allSensorsOff() {
+		if (!back.get() && !backMid.get() && !mid.get() && !frontMid.get() && !front.get()) {
+			return true;
+		}
+
+		return false;
+		// check if all sensors are on
+	}
+
+	public void shoot() {
+		// allShiftDown();
+		indexMidMotor.set(1);
+		if (ballUnbalancedSide == IntakeDirection.FRONT) {
+			indexBackMotor.set(-1);
+		} else {
+			indexFrontMotor.set(1);
+		}
+		if (frontMid.get() && front.get()) {
+			indexFrontMotor.set(0);
+			ballUnbalancedSide = IntakeDirection.FRONT;
+		} else if (backMid.get() && back.get()) {
+			indexBackMotor.set(0);
+			ballUnbalancedSide = IntakeDirection.BACK;
+		} else if (allSensorsOff()) {
+			stopAll();
+			ballUnbalancedSide = IntakeDirection.NONE;
+		}
+	}
+
+//Switch ball positions
+
+	public enum IntakeDirection {
+		FRONT, BACK, BOTH, NONE;
+	}
+
+	public void intake(int num, IntakeDirection dir) {
+		switch (num) {
+		case 1:
+			if (dir == IntakeDirection.FRONT) {
+				indexFrontMotor.set(1);
+			} else if (dir == IntakeDirection.BACK) {
+				indexBackMotor.set(-1);
+			}
+			if (!mid.get()) {
+				stopAll();
+				numBalls++;
+			}
+			break;
+		case 2:
+			indexFrontMotor.set(1);
+			indexBackMotor.set(-1);
+			if (!frontMid.get()) {
+				stopAll();
+				ballUnbalancedSide = IntakeDirection.BACK;
+				numBalls++;
+			} else if (!backMid.get()) {
+				stopAll();
+				ballUnbalancedSide = IntakeDirection.FRONT;
+				numBalls++;

```

Comment:
Hrm.  Presumably the IntakeBalls command should continue running until one ball is added to the indexer system.  That implies the IntakeBalls command needs to know when the indexer subsystem is done intaking a ball so it can then return `true` from the `isFinished()` method.  The IntakeBalls command also may want to be interruptible so a driver can abort intaking a ball if they want to start shooting balls instead.  Similarly, the driver should be able to manually abort/interrupt intaking a ball.

Reply:
This is complex enough it's worth diagramming out the various possible states, the allowable changes between those states, and what are the start & end conditions for those state changes.